spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,201 @@
+private[cassandra] class TableScanRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                               userSpecifiedSchema: Option[StructType], sqlcontext: SQLContext)
+  extends BaseRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlcontext) with TableScan {
+  override def buildScan(): RDD[Row] = super.buildScan()
+}
+
+/** Table scan relation implements [[BaseRelation]], [[InsertableRelation]] and [[PrunedScan]] */
+private[cassandra] class PrunedScanRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                                userSpecifiedSchema: Option[StructType], sqlcontext: SQLContext)
+  extends BaseRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlcontext) with PrunedScan {
+  override def buildScan(requiredColumns: Array[String]): RDD[Row] = {
+    val transformer = new RddFilterSelectPdTrf(requiredColumns, columnNameByLowercase, Seq.empty)
+    (transformer.maybeSelect apply baseRdd).asInstanceOf[RDD[Row]]
+  }
+}
+
+/** Table scan relation implements [[BaseRelation]], [[InsertableRelation]] and [[PrunedFilteredScan]] */
+private[cassandra] class PrunedFilteredScanRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                                        userSpecifiedSchema: Option[StructType], sqlcontext: SQLContext)
+  extends BaseRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlcontext) with PrunedFilteredScan {
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    val pushDown = new FilterPushDown(filters, tableDef)
+    val pushdownFilters = pushDown.toPushDown
+    val preservedFilters = pushDown.toPreserve
+
+    logInfo(s"pushdown filters: ${pushdownFilters.toString()}")
+
+    // a filter combining all other filters
+    def rowFilter(row: Row): Boolean = !preservedFilters.map(translateFilter(_)(row)).contains(false)
+
+    val transformer = new RddFilterSelectPdTrf(requiredColumns, columnNameByLowercase, pushdownFilters)
+    transformer.transform(baseRdd).asInstanceOf[RDD[Row]].filter(rowFilter)
+  }
+
+  /** Evaluate filter by column value from the row */
+  private[this] def translateFilter(filter: Filter): Row => Boolean = filter match {
+    case EqualTo(column, v) => (a: Row)            => compareColumnValue(column, a, v) == 0
+    case LessThan(column, v) => (a: Row)           => compareColumnValue(column, a, v) < 0
+    case LessThanOrEqual(column, v) => (a: Row)    => compareColumnValue(column, a, v) <= 0
+    case GreaterThan(column, v) => (a: Row)        => compareColumnValue(column, a, v) > 0
+    case GreaterThanOrEqual(column, v) => (a: Row) => compareColumnValue(column, a, v) >= 0
+    case In(column, values) => (a: Row) => val (value, dataType) = getSchemaData(column, a)
+      values.map(_.asInstanceOf[dataType.JvmType]).toSet.contains(value)
```

Comment:
Mapping values just to change the type does nothing when we just want to provide the value exists in the set.